File: matlab/functions/preprocessClouds.m
function arrLiDAR_ordered = preprocessClouds(configpath)
%UNTITLED2 Summary of this function goes here
%   Detailed explanation goes here
    %txt = fileread("C:\Users\agori\Documents\MATLAB\3D-Sea-Ice\matlab\config\p3_config.json");
    logInfo(sprintf("Loading preprocessing configuration file %s",configpath));
    tStage = tic;
    tTotal = tic;
    txt = fileread(configpath);
    params = jsondecode(txt);
    outputFolder = params.outputFolder;
    if ~exist(outputFolder, 'dir')
        mkdir(outputFolder);
    end
    pathname = params.preprocessing.pathname;
    gridStep = params.preprocessing.gridStep;
    x_roi = decodeJsonROI(params.preprocessing.roi.x);
    y_roi = decodeJsonROI(params.preprocessing.roi.y);
    z_roi = decodeJsonROI(params.preprocessing.roi.z);
    
    post_x_roi = decodeJsonROI(params.preprocessing.postFloorROI.x);
    post_y_roi = decodeJsonROI(params.preprocessing.postFloorROI.y);
    post_z_roi = decodeJsonROI(params.preprocessing.postFloorROI.z);
    postROI=[post_x_roi, post_y_roi, post_z_roi];

    nclouds = params.preprocessing.nclouds;
    random_seed = params.preprocessing.seed;
    startCloud = params.preprocessing.startCloud;
    scanOrder = params.preprocessing.scanOrder;
    save_arrLiDAR = params.preprocessing.save_arrLiDAR;
    roi = [x_roi, y_roi, z_roi];
    rng(random_seed, "twister");
    
    logInfo(sprintf("Configuration loaded successfully (+%.3fs)", toc(tStage)));
    
    
    logInfo(sprintf("Reading ROS bag files from %s",pathname));
    tStage = tic;
    
    [all_clouds, fnames] = rosGetRaw(pathname);
    %[minsz, testout] = getMinSz(all_clouds, fnames);
    
    logInfo(sprintf("ROS bag parsing complete (+%.3fs)", toc(tStage)));
    
    
    msg = sprintf([ ...
        'Performing cropMergeDownsample:\n' ...
        '    ROI crop:\n' ...
        '        X = %s\n' ...
        '        Y = %s\n' ...
        '        Z = %s\n' ...
        '    Voxel grid size = %.3f m\n' ...
        '    Frame merge count = %d\n' ...
        '    Frame range = [%d, %d]'], ...
        mat2str(x_roi), ...
        mat2str(y_roi), ...
        mat2str(z_roi), ...
        gridStep, ...
        nclouds, ...
        startCloud, ...
        startCloud + nclouds - 1);
    
    logInfo(msg);
    tStage = tic;
    
    arrLiDAR = cropMergeDownsample(roi, nclouds, gridStep, all_clouds, startCloud);
    
    logInfo(sprintf("Point cloud preprocessing complete (+%.3fs)", toc(tStage)));
    
    
    logInfo(sprintf("Reordering point clouds using fieldnotes %s", scanOrder));
    tStage = tic;
    
    arrLiDAR_ordered = reorderPointCloud(arrLiDAR, scanOrder);
    
    logInfo(sprintf("Point cloud reordering complete (+%.3fs)", toc(tStage)));
    
    if save_arrLiDAR
        arrLiDARfolder=strcat(outputFolder,params.preprocessing.save_arrLiDARfolder);
        saveClouds(arrLiDAR_ordered,arrLiDARfolder,params.preprocessing.save_arrLiDARname);
    end

    alnFilename = strcat(outputFolder, "floorTforms.aln");
    logInfo(sprintf("Performing floor removal saving to %s", alnFilename));
    tStage = tic;
    %Floor removal
    arrLiDAR=arrLiDAR_ordered;
    croppedClouds=arrLiDAR;
    croppedRotated=arrLiDAR;
    floorTforms = cell(8,1);
    %remove floor points for each arrCloudTrans
    for k=1:8
        [floorAlign,tform]=alignFloorTform(arrLiDAR_ordered(k));
        floorTforms{k}=tform;
        invTform = invert(tform);
        %crop floor out
        [sortedPOI,arrIndex,palletIndex,floorIndex,surfIndex] = normsAnalysis(floorAlign);
        croppedClouds(k)=cropCloud(floorAlign,[-inf,inf],[-inf,inf],[sortedPOI(floorIndex+1)+0.02,inf]);
        croppedRotated(k)=pctransform(croppedClouds(k),invTform);
    end
    
    writeALNtransformRigid(alnFilename,floorTforms);
    
    logInfo(sprintf("Floor removal completed and saved (+%.3fs)", toc(tStage)));
   

    msg = sprintf([ ...
        'Performing post floor removal crop:\n' ...
        '    ROI crop:\n' ...
        '        X = %s\n' ...
        '        Y = %s\n' ...
        '        Z = %s\n' ], ...
        mat2str(post_x_roi), ...
        mat2str(post_y_roi), ...
        mat2str(post_z_roi));
    
    logInfo(msg);
    tStage = tic;
    arrLiDAR_rotated=cropClouds(croppedRotated,postROI);
    logInfo(sprintf("Post floor crop complete (+%.3fs)", toc(tStage)));

    
    logInfo(sprintf("Extracting coarse alignment from %s", params.preprocessing.coarseALN));
    tStage = tic;
    arrTformsRigid=extractALNtransformRigid(params.preprocessing.coarseALN);
    arrCoarse = applyRStforms(arrLiDAR_rotated,arrTformsRigid);
    BeforeICPpath=strcat(outputFolder,params.preprocessing.beforeICPpath);
    
    logInfo(sprintf("Saving coarse aligned clouds to  %s", BeforeICPpath));
    saveClouds(arrCoarse,BeforeICPpath,params.preprocessing.fname);
    matFilename = strcat(BeforeICPpath, "beforeICP.mat");
    save(matFilename,"arrCoarse");
    logInfo(sprintf("Coarse alignment completed and saved (+%.3fs)", toc(tStage)));
    
    totalTime = toc(tTotal);
    hours = floor(totalTime / 3600);
    minutes = floor(mod(totalTime, 3600) / 60);
    seconds = mod(totalTime, 60);
    
    logInfo(sprintf("Pipeline completed successfully. Total runtime = %02dh:%02dm:%05.2fs",hours, minutes, seconds));
end

File: python-3DRecon/preprocessClouds.py
from extractclouds import (
    ros_get_raw,
    get_min_sz,
)
from pathlib import Path
import numpy as np
import open3d as o3d
import pyvista as pv
from pcdisplay import show_pointcloud_intensity, plotter_pcdisplay
from rosbags.highlevel import AnyReader
import json
import pandas as pd
import pctools
import pcalign
from roiutils import decode_roi

#o3d.utility.set_verbosity_level(o3d.utility.VerbosityLevel.Error)
with open("p3_config.json", "r") as f:
    config = json.load(f)

pathname = config["preprocessing"]["pathname"]

all_clouds, fnames = ros_get_raw(pathname)
minsz, testout = get_min_sz(all_clouds,fnames)

cloud = pv.PolyData(testout["xyz"])
cloud["intensity"] = testout["intensity"]
cloud.save("testout.vtp")
del cloud
#plotter1 = plotter_pcdisplay(cloud)

#Read ROI
roi = decode_roi(config["preprocessing"]["roi"])

arr_lidar = pctools.crop_merge_downsample(
    roi=roi,
    nclouds=50,
    ds_grid_step=0.01,
    all_clouds=all_clouds,
    start_cloud=9,
)

#pcd = arr_lidar[0]

#xyz = pcd.point["positions"].numpy()
#intensity = pcd.point["intensity"].numpy().flatten()

#cloudOut = pv.PolyData(xyz)
#cloudOut["intensity"] = intensity

#show_pointcloud_intensity(cloudOut)
#plotter2 = plotter_pcdisplay(cloudOut)

arr_ordered = pctools.reorder_point_cloud(arr_lidar, config["preprocessing"]["scanOrder"])


arr_lidar = arr_ordered

cropped_clouds = [None] * len(arr_lidar)
cropped_rotated = [None] * len(arr_lidar)
floor_tforms = [None] * len(arr_lidar)
#Floor adjustment 
for k in range(len(arr_lidar)):
    # Align floor to horizontal
    floor_align, tform = pctools.align_floor_tform(arr_lidar[k])

    floor_tforms[k] = tform

    # Inverse transform
    inv_tform = np.linalg.inv(tform)

    # Analyse normal distribution
    sorted_poi, arr_index, pallet_index, floor_index, surf_index = pctools.norms_analysis(
        floor_align,
        show_graph=False
    )

    z_cutoff = sorted_poi[floor_index + 1] + 0.02

    # Crop floor out in aligned frame
    x_roi=[-np.inf, np.inf]
    y_roi=[-np.inf, np.inf]
    z_roi=[z_cutoff, np.inf]
    roi = [x_roi, y_roi, z_roi]
    cropped_clouds[k] = pctools.crop_o3d_t(
        floor_align, roi
    )

    # Transform cropped cloud back to original orientation
    cropped_rotated[k] = cropped_clouds[k].clone()
    cropped_rotated[k].transform(inv_tform)
    del floor_align

#Write floortforms to file
before_icp_path = config["outputFolder"] + config["preprocessing"]["beforeICPpath"]
scan_labels = [config["preprocessing"]["fname"] + ch for ch in "ABCDEFGH"[:len(floor_tforms)]]
pcalign.write_aln(before_icp_path + "floor_tforms.aln", floor_tforms, scan_labels)


#Perform alignment
coarse_tforms, _ = pcalign.read_aln(config["preprocessing"]["coarseALN"])
arr_coarse = pcalign.apply_rs_tforms(cropped_rotated, coarse_tforms)


pcalign.save_clouds(arr_coarse, before_icp_path, config["preprocessing"]["fname"])
#show_pointcloud_intensity(cropped_rotated[0])

del cropped_clouds, cropped_rotated, floor_tforms, arr_lidar
import gc; gc.collect()